ESQL: Serialize the source in expressions

bpintea opened the pull request:
When de-/serializing expressions, the source is currently omitted. This
is however required to correctly issue Warning headers. This adds
serialisation of the source.

The entire query is now added to the EsqlConfiguration and serialized
entirely and just once, part of it. This allows each expression to only
serialise source's length, since this, together with the Location, allows
reconstructing the source from config. The query string is also 
compressed, if exceeding a certain size (5K chars).

The Warning messages are also modified to always include the location
of the query whose evaluation generated an exception. This is required
because same exception can be generated from different query
statements and the Warning caching code deduplicates the headers.

bpintea committed:
Serialize also the source in expressions

When de-/serializing expressions, the source is currently omitted. This
is however required to correctly issue Warning headers. This adds
serialization of the source.

The Warning messages are modified to always include the location of the
query whose evaluation generated an exception. This is required because
same exception can be generated from different query statements and the
Warning caching code deduplicates the headers.

bpintea committed:
Send the whole source just once, part of the cfg

This removes the text representation of the source for each
expression's serialization and replaces that with its lenght only.

The entire query is now added to the EsqlConfiguration and serialized
entirely and just once part of it. It can then be compressed further, if
needed (todo).

bpintea committed:
Minor adjustments

Remove commented code, rearange a method.

bpintea committed:
Merge branch 'main' into esql/serialize_source_in_expressions

Commit:
Update docs/changelog/99956.yaml

elasticsearchmachine commented:
Hi @bpintea, I've created a changelog YAML for you.

bpintea commented:
@elasticsearchmachine run elasticsearch-ci/bwc

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

elasticsearchmachine commented:
Pinging @elastic/elasticsearch-esql (:Query Languages/ES|QL)

ChrisHegarty commented:
The serialization/deserialization of the source seems reasonable to me. I noticed that we just do it in some places - there are a lot `Source.EMPTY` remaining in PlanNamedTypes. Is it the case that the source is not needed or applicable there? I'm wondering if we should make source and optional serialization element everywhere - allowing to add a non-null source later in other places?

bpintea commented:
> I noticed that we just do it in some places - there are a lot Source.EMPTY remaining in PlanNamedTypes. Is it the case that the source is not needed or applicable there?

The source is currently serialised only for the expressions that can generate `null` (and are to generate a [Warning](https://github.com/elastic/elasticsearch/issues/98743) too).

> I'm wondering if we should make source and optional serialization element everywhere - allowing to add a non-null source later in other places?

Also IMO it would make sense. I'll extend the PR.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/EsqlTestUtils.java:
```diff
@@ -93,7 +98,21 @@ public static EsqlConfiguration configuration(QueryPragmas pragmas) {
+    }
+
+    public static EsqlConfiguration configuration(String query) {
+        return new EsqlConfiguration(
+            DateUtils.UTC,
+            Locale.US,
+            null,
+            null,
+            new QueryPragmas(Settings.EMPTY),
+            EsqlPlugin.QUERY_RESULT_TRUNCATION_MAX_SIZE.getDefault(Settings.EMPTY),
+            EsqlPlugin.QUERY_RESULT_TRUNCATION_DEFAULT_SIZE.getDefault(Settings.EMPTY),
+            query
```
super-duper nit: to avoid copy-pasting, we could factor out the common parts from this and above into a `private static EsqlConfiguration configuration(QueryPragmas pragmas, String query)`.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/AnalyzerTestUtils.java:
```diff
@@ -44,6 +47,10 @@ public static Analyzer analyzer(IndexResolution indexResolution, Verifier verifi
         );
     }
 
+    public static Analyzer analyzer(EsqlConfiguration config, IndexResolution indexResolution, Verifier verifier) {
```
nit: argument order breaks the telescope pattern; would be slightly more expected if this was
```suggestion
    public static Analyzer analyzer(IndexResolution indexResolution, Verifier verifier, EsqlConfiguration config) {
```

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamInputTests.java:
```diff
@@ -86,4 +96,54 @@ List<Long> randomLongsListOfSize(int size) {
     }
+
+    public void testSourceSerialization() {
+        Function<String, String> queryFn = d -> d
```
super-duper nit: since as a reader I don't have much context when I start reading this test, using an expressive lambda argument name could help a little.
```suggestion
        Function<String, String> queryFn = lineDelimiter -> lineDelimiter
```

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/Warnings.java:
```diff
@@ -10,33 +10,37 @@
 
     public Warnings(Source source) {
-        this.source = source;
+        location = format("Line {}:{}: ", source.source().getLineNumber(), source.source().getColumnNumber());
+        first = format(
+            null,
+            "{}evaluation of [{}] failed, treating result as null. Only first {} failures recorded.",
+            location,
+            source.text(),
+            MAX_ADDED_WARNINGS
```
Nit: Hm, the impact of that may be negligible, but this refactoring means that we eagerly create the formatted `first` warning message and the `location` string whenever a `Warnings` instance is created. May be slightly better to keep the logic lazy as it was.

bpintea replied:
> this refactoring means that we eagerly create the formatted first warning message and the location string whenever a Warnings instance is created.

Indeed.

> May be slightly better to keep the logic lazy as it was.

Why? Memory-consumption concerns?
The change adds location information to the exception (registering) line as well. So the `location` needs to be built at construction time anyways (or at least the line/column recorded). The `first` could be deferred, but otoh, that'd be in the data handling path (not that it would matter much, though).

alex-spies replied:
Preface: please treat this comment as optional, it's not a big deal either way :)

As to why: Less about memory consumption, more that we're adding multiple calls to `format` now whenever an evaluator is created, as evaluators come with a `Warnings` instance; for an expression/function-heavy query, this might be (very, very, very slightly) affecting performance. For the happy path where no warning is issued, neither the `first` string nor the `location` string need to be built from the source; in the unhappy path, of course, the location is needed for every warning and then caching the location string is faster, but path should be hit less often.

Again, probably negligible so simplicity is more important than performance concerns. Just pointed it out because the previous approach (which only kept track of the source) looked simple enough.

alex-spies reviewed (commented):
Generally LGTM and could be merged as-is in the current state; there's one potential off-by-one bug in the source de/serialization that would be good to confirm if I spotted that correctly, though.

bpintea committed:
Merge branch 'main' into esql/serialize_source_in_expressions

bpintea committed:
Address review comments

Improve the handling of column, add comments, refactor a method.

bpintea committed:
Make all source optional serializable

Enable serialization of source for all plan types.

alex-spies reviewed (approved):
LGTM!

bpintea committed:
GZip the query if longr than a threashold

This adds compression to the query string, if this gets overly long (5K
chars).

bpintea committed:
Merge branch 'main' into esql/serialize_source_in_expressions

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamInput.java:
```diff
@@ -101,6 +105,43 @@ public PhysicalPlan readPhysicalPlanNode() throws IOException {
+    }
+
+    private static String sourceText(String query, int line, int column, int length) {
+        if (line <= 0 || column <= 0 || query.isEmpty()) {
```
Thanks, addressed by serialising the 1-based column, since this is going to be reported in any error as well.

bpintea commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/ints.csv-spec:
```diff
@@ -472,3 +472,41 @@ ROW deg = [90, 180, 270]
+null
+;
+
+// waiting on https://github.com/elastic/elasticsearch/pull/99926
```
Removed, thanks.

bpintea commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamInput.java:
```diff
@@ -101,6 +105,43 @@ public PhysicalPlan readPhysicalPlanNode() throws IOException {
+                "location [@" + line + ":" + column + "] and length [" + length + "] overrun query size [" + query.length() + "]"
+            );
+        }
+        return query.substring(offset, offset + length);
```
This is [now](https://github.com/elastic/elasticsearch/pull/99956/commits/6ad8cf852ce34f4fa9153499a43b8212882973f0) also compressed.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlConfiguration.java:
```diff
@@ -13,13 +13,23 @@
 
 public class EsqlConfiguration extends Configuration implements Writeable {
+
+    static final int QUERY_COMPRESS_THRASHOLD_CHARS = KB.toIntBytes(5);
```
typo:
```suggestion
    static final int QUERY_COMPRESS_THRESHOLD_CHARS = KB.toIntBytes(5);
```

alex-spies reviewed (approved):
LGTM!

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlConfiguration.java:
```diff
@@ -81,20 +97,52 @@ public Locale locale() {
+        return query;
+    }
+
+    private static void writeQuery(StreamOutput out, String query) throws IOException {
+        if (query.length() > QUERY_COMPRESS_THRASHOLD_CHARS) { // compare on chars to avoid UTF-8 encoding unless actually required
+            out.writeBoolean(true);
+            var baos = new ByteArrayOutputStream();
+            try (var gzip = new GZIPOutputStream(baos)) {
+                gzip.write(query.getBytes(StandardCharsets.UTF_8));
+            }
+            out.writeByteArray(baos.toByteArray());
+        } else {
+            out.writeBoolean(false);
+            out.writeString(query);
+        }
+    }
+
+    private static String readQuery(StreamInput in) throws IOException {
+        boolean compressed = in.readBoolean();
+        if (compressed) {
+            byte[] bytes = in.readByteArray();
+            var bais = new ByteArrayInputStream(bytes);
+            try (var gzip = new GZIPInputStream(bais)) {
+                return new String(gzip.readAllBytes(), StandardCharsets.UTF_8);
+            }
+        } else {
+            return in.readString();
+        }
+    }
+
```
Have you checked `org.elasticsearch.common.compress`?
It relies only on the Deflate/Inflate compressor instead of the whole gzip - in our case since we're doing string compression, we don't expect things to be too large anyway.

bpintea replied:
I have, but seemed heavier than the need here. I switched over though. Thanks.

costin reviewed (approved):
LGTM

bpintea committed:
Merge branch 'main' into esql/serialize_source_in_expressions

bpintea committed:
Adress review points

- typo in var name
- switch from GZip to ES's compression
